### PR TITLE
Addition of classifier-free guidance notebook 

### PR DESCRIPTION
**Pull request by virginiafdez**
### Description
Addition of classifier-free guidance notebook for the use of CFG on DDPMs. 

### Checks
- [X] Avoid including large-size files in the PR.
- [X] Clean up long text outputs from code cells in the notebook.
- [X] For security purposes, please check the contents and remove any sensitive info such as user names and private key.
- [X] Ensure (1) hyperlinks and markdown anchors are working (2) use relative paths for tutorial repo files (3) put figure and graphs in the `./figure` folder
- [X] Notebook runs automatically `./runner.sh -t <path to .ipynb file>`

### CODE REVIEW

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+   "id": "63d95da6",
+   "metadata": {},
+   "source": [
+    "# Classifier-free Guidance\n",
+    "\n",
+    "This tutorial illustrates how to use MONAI for training a denoising diffusion probabilistic model (DDPM)[1] to create synthetic 2D images using the classifier-free guidance technique [2] to perform conditioning.\n",
+    "\n",
+    "\n",
+    "[1] - Ho et al. \"Denoising Diffusion Probabilistic Models\" https://arxiv.org/abs/2006.11239\n",
+    "[2] - Ho and Salimans \"Classifier-Free Diffusion Guidance\" https://arxiv.org/abs/2207.12598\n",
```
Comment (ericspod) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:31]:
```suggestion
    "- [1] Ho et al. \"Denoising Diffusion Probabilistic Models\" https://arxiv.org/abs/2006.11239\n",
    "- [2] Ho and Salimans \"Classifier-Free Diffusion Guidance\" https://arxiv.org/abs/2207.12598\n",
```
Slightly better layout. It would also help to explain here what classifier-free guidance is and what it does for you briefly in one sentence.

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb [hunk 2]
```diff
@@ -0,0 +1,959 @@
...
.../+c8nx7/4xS/KmK997Wtybs2aNclxrzLWc8sttyTHVVNuM7Of//znyXGv4nhoaEjOqe+TV3FZU1Mj50pdfY2pxZMdACA8kh0AIDySHQAgPJIdACA8kh0AIDySHQAgvIpbelDppbmT7Jt9jNxjUp81OjoqYw4ePCjn7r333uT4RRddVNyOmV+u7R3v7t27k+O33367jPGOd+PGjcnxDRs2yJgXXnghOe6V9r/88styTi1zOPHEE2VMQ0ODnNu1a1dyXDXRNjPbsmVLcvz555+XMb29vXLuW9/6VnLca7RcVVUl5xYvXpwcv+OOO2TM73//++S4d++NjIzIObV/XiNoT85vAaYPT3YAgPBIdgCA8Eh2AIDwSHYAgPBIdgCA8OYUJllStGrVKjkXsSpJVRR6TY5VE+achtPenFc95lUh7ty5Mzl+0kknyRjF24d9+/bJua1btybHverEtrY2OdfU1JQc9xo3K3v37pVzXpXf+eefnxz3rq3XsFjth1eVqppYz58/X8Z49/JZZ52VHL/55ptljHeOvIbKynXXXZccf+qpp2RMfX29nFPnL/f7maPSq81nKu834j082QEAwiPZAQDCI9kBAMIj2QEAwiPZAQDCI9kBAMKruEbQla6cjaC90nDlyiuvlHMrVqxIjnvHpPZdLbMwM3vllVfk3Nq1a5PjR44ckTEtLS1ybuHChclxrxRZlaB3dHTImHPOOUfOtba2JsefeOIJGXPo0CE5t3r16uT4Bz/4QRmjGi339PTImJUrV8q55ubm5Pjg4KCM8cr+Fa/59he+8IXk+OOPPy5jGhsb5Vx1dXVy3Gss7WEZwczCkx0AIDySHQAgPJIdACA8kh0AIDySHQAgPJIdACC8SS89iPhmA4/q6l9VVSVjVDm+94YAjzrnXin+5s2b5Zza95xu9V7Z9fbt2+XcgQMHkuNeSbtXGq668B8+fFjGvPPOO8lx7w0B3jX8yU9+khz3rsXHP/5xObds2bLkuLcURR2vt7xg/fr1cq67uzs57i2Z8N6Mos6fdx9deumlyfFTTz1VxrS3txe9D9797y2xUd9PliRUJp7sAADhkewAAOGR7AAA4ZHsAADhkewAAOHN6kbQOVVTOTFeJZ9X3ZnThFlVO3pUxaWZbtDsHdOZZ54p5959993kuKpANDN77bXX5Jw6RyeccIKMUc2HvepEr9K
```
Comment (ericspod) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:721]:
```suggestion
    "    progress_bar = tqdm(train_loader, total=len(train_loader), ncols=70, disable=True)\n",
    "    for batch in progress_bar:\n",
```
We can remove the enumerate if we're not using the step value at all.

Comment (ericspod) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:752]:
```suggestion
    "        for batch in val_loader:\n",
    "            images = batch[\"image\"].to(device)\n",
```

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb [hunk 3]
```diff
@@ -0,0 +1,959 @@
+{
```
Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>Would be better to put [2] on a separate line.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    print(f"batch shape: {check_data['image'].shape}")
</pre><p>Could you please add a shape check for the class as well?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>It seems that the loss is not decreasing. Is this expected?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='23'/>

Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>I believe the definition for&nbsp;<code>guidance_scale</code>&nbsp;is in Reference [2]. It would be helpful to add a brief explanation here to direct users to obtain more information from [2].</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/Project-MONAI/tutorials/pull/1867/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb (hunk 2 above)
Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:721]: To resolve the pep8 error: https://github.com/Project-MONAI/tutorials/actions/runs/11573173298/job/32214701041?pr=1867#step:7:304
```suggestion
    "    for _step, batch in progress_bar:\n",
```

Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:751]:
```suggestion
    "        for _step, batch in enumerate(val_loader):\n",
```

**Review by KumoLiu — APPROVED** [review]
Thanks for the PR!
Overall looks good to me. Just leave several minor comments inline.

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb (hunk 3 above)
Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>Done!</p>

Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>Done</p>

Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>I think the issue here is that the LDM loss goes down at the start (somehow not captured because it probably goes down quickly within the first epoch). Other than that, it's a noisy loss, which causes it to look like this sometimes even when the model's learning. </p>

Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:1]: <p>Done. </p>

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb (hunk 2 above)
Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:721]: Done for the val_loader. For the train one you are using step.

Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:721]: I just removed step. I was wrong in a previous comment, it's actually not needed

Comment (virginiafdez) [reply generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:751]: I removed it, not needed.

File: generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb [hunk 4]
```diff
@@ -0,0 +1,880 @@
...
+    "jupyter": {
+     "outputs_hidden": false
+    }
+   },
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "/tmp/ipykernel_4172533/2226485125.py:15: UserWarning: To copy construct from a tensor, it is recommended to use sourceTensor.clone().detach() or sourceTensor.clone().detach().requires_grad_(True), rather than torch.tensor(sourceTensor).\n",
```
Comment (KumoLiu) [inline generation/classifier_free_guidance/2d_ddpm_classifier_free_guidance_tutorial.ipynb:373]: Would be better to remove these warning to make the output clean.
